aptos-move/aptos-vm/src/aptos_vm.rs

```diff
@@ -408,13 +424,72 @@ impl AptosVM {
         self.execute_module_initialization(
             &mut session,
             gas_status,
-            modules,
+            &self.deserialize_module_bundle(modules)?,
             &[txn_data.sender()],
         )?;
 
         self.success_transaction_cleanup(session, gas_status, txn_data, log_context)
     }
 
+    /// Resolve a pending code publish request registered via the NativeCodeContext.
+    fn resolve_pending_code_publish<S: MoveResolverExt>(
+        &self,
+        session: &mut SessionExt<S>,
+        gas_status: &mut GasStatus,
+    ) -> VMResult<()> {
+        if let Some(PublishRequest {
+            destination,
+            bundle,
+            expected_modules,
+            check_compat,
+        }) = NativeCodeContext::extract_publish_request(session)
+        {
+            // TODO: unfortunately we need to deserialize the entire bundle here to handle
+            // `init_module` and verify some deployment conditions, while the VM need to do
+            // the deserialization again. Consider adding an API to MoveVM which allows to
+            // directly pass CompiledModule.
+            let modules = self.deserialize_module_bundle(&bundle)?;
+
+            // Validate the module bundle
+            self.validate_publish_request(&modules, expected_modules)?;
+
+            // Publish the bundle
+            if check_compat {
```

Comment:
It seems risky here to allow the ability to upgrade and bypass compatibility check. Do we need this right away? If not, may be we should omit for simplicity + less risk.

Reply:
This for initial experimentation. A package marked like this cannot be consumed as a dependency. See also design doc.

Reply:
What do you mean by initial experimentation? I'm not sure which part of the design doc talks about this but I think we should not allow any bypass of the compatibility check when upgrading a module.

Reply:
User's initial experimentation with publishing their package. Its called the zero-version in the doc, right now its not longer a version, but expressed in the `no-compat` policy.

This is a standard concept known from semantic versioning. There, any 0.x.y version is not bound to any compatibility rules. This allows one too quickly evolve initial draft code. In turn, a consumer of the package is clearly signaled that he cannot expect any compatibility guarantees. Or we may even disallow consuming such a package in the first place, but this happens then on cli/compiler side.

This has been in the design since the very beginning (even earlier docs with semver), and requested by reviewers. I much prefer to keep this in as I do not see any practical danger besides someone shooting in their own food with this.

Reply:
I'm quite sure I understand here. Are you saying users can always bypass the compatibility check for their first upgrade (of version 0 to version 1)?